### ISSUE TITLE
Window Outlines are Blurry on High DPI Setups

### ISSUE THREAD

**Issue by Infinality** [paulmcauley/klassy#155]
On my 150% scaled setup, 1px window outlines appear blurry and **thicker** than 1px.  What I suspect is happening is that the 1px border is being calculated on a standard 100% basis and then when the scaling is at 150%, you're actually getting 1px scaled up to 1.5 pixels.  Ideally, klassy should be aware that it's running under a certain scaling (125%, 150%, 200%, etc.) so that 1px is literally rendered as 1px.   Or as a workaround, outline width pixel values less than 1 should be allowed.  For instance:  If I were to set the value at 0.66, it would counteract the artificially inflated 1.5px value back to 1px.

**Comment by paulmcauley** [comment]
The outline thickness also includes a 0.5 pixel overlap to work around fractional scaling errors. To fix properly requires a kwin/KDecoration2 API update.

For the second point, does this also happen in Breeze application style, as this is most likely inherited.

**Comment by Infinality** [comment]
You're right:  The second thing _does_ happen with Breeze application style too.  It's interesting (but not surprising) that most of these issues are actually upstream issues.  I figured after ~10 years of development, the Linux people would have have high DPI down to a science, but it's still a glitchy mess.   ¯\_(ツ)_/¯

Does it make sense to allow values lower than 1.00 as a workaround or will that produce obviously bad results in all cases?

**Comment by paulmcauley** [comment]
To do it better, at a minimum the KDecoration2 API needs to give the decoration access to DevicePixelRatioF.

At 150% scaling it will give 1 pixel outline, with an additional 0.5 px overlap, if you set the thickness to 1.0.

PS: are you the guy behind the Freetype patches? I was thinking an alternative to writing my own pixel-snapping routines as I have done here, a lot of the icons could be specified as a font and sub-pixel snapped if some of the same techniques from Freetype were used. I am not that familiar with Freetype though.

**Comment by Infinality** [comment]
Yes, I am, but if you can believe it, even though I've done lots of coding I've never actually implemented any code that leverages freetype.  And, I knew very little about it (or fonts) when I starting playing with the code and wrote those patches.

What you're suggesting sounds plausible, but the advantage of doing so may or may not be worth the extra complexity.

**Comment by Infinality** [comment]
Here's another thing I found.  I'm guessing it's probably the same thing where the API needs to be updated:

This is a transparent console window with blur applied.  You can see the klassy blue window border at the bottom, but if you look between the border and the dark console area, you can see an _exact_ 1px horizontal line between them.  It appears to be 100% transparent, but blurred, of whatever is underneath this window.  Strangely, it doesn't happen all the time, just sometimes.

![1px-issue](https://github.com/user-attachments/assets/1ebdb477-9660-4b69-b4b2-036955afd9dd)

**Comment by AlexFolland** [comment]
I set my window outlines to exactly 1.00 pixel thickness and 100% opacity, but it is still partially transparent, blending with the nearby pixels.  Something seems wrong with the calculation of its color and/or position.

**Comment by Infinality** [comment]
My guess is that this also is a KDecoration2 API issue like the one I reported.  I'm continually surprised that we're ~ 15 years into "High DPI" technology and all of these codebases & APIs are riddled with problems.  ¯\\_(ツ)_/¯

**Comment by paulmcauley** [comment]
> I set my window outlines to exactly 1.00 pixel thickness and 100% opacity, but it is still partially transparent, blending with the nearby pixels. Something seems wrong with the calculation of its color and/or position.

A 0.5 pixel overlap is added to account for fractional scaling errors. If you are at 100% scaling then set to 1.5 for a sharp 1px outline

**Comment by AlexFolland** [comment]
OK, I tested that and it is indeed true.  That sounds like it is the cause of this reported issue, at least for me.  The compensation for a potential issue seems to have introduced an actual issue, where the configured value doesn't match the value that's used, leading to needing a workaround.  Can we expect the compensation to be removed, to fix this reported issue?

**Comment by Infinality** [comment]
It appears the issue is still present (or maybe worse) on the 6.3 version.  The outline thickness here is set to 1.00, yet it apperas to be 3-4 pixels thick.  My display scaling is set at 1.5.  It appears that the _outline itself_ may be getting blurred.


![Image](https://github.com/user-attachments/assets/876d4e85-d38e-44e1-aee0-e5a722210185)

**Comment by Infinality** [comment]
Also, the behavior is strange.  If I set the opacity of the outline to 100% transparent, there is no outline against a black background, but there is against a white one.  However, against the white background, the bottom and right outline is missing on the active window.  The behavior I'd expect is that in all cases, there would be a 1px outline on all sides when the opacity is greater than transparent, and no outline whatsoever when the opacity is 0.

![Image](https://github.com/user-attachments/assets/d5464223-337d-4cd2-adc9-88aec10d449e)

![Image](https://github.com/user-attachments/assets/7667ba40-108c-4d8f-b558-f1130e2351ab)

**Comment by dezdem** [comment]
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/d8337971-bbd8-4160-bbe6-b5024e44dbe3" />
100% scaling here, and couldn't get a sharp corners. Is this the same culprit here?
